docs/design/generics/terminology.md

```diff
@@ -349,27 +452,24 @@ instead of associated entity.
 
 An _impl_ is an implementation of an interface for a specific type, called the
 _implementing type_. It is the place where the function bodies are defined,
-values for associated types, etc. are given. Implementations are needed for
+values for associated constants, etc. are given. Implementations are needed for
 [nominal interfaces](#nominal-interfaces);
 [structural interfaces](#structural-interfaces) and
 [named constraints](#named-constraints) define conformance implicitly instead of
-by requiring an impl to be defined. In can still make sense to implement a named
-constraint as a way to implement all of the interfaces it requires.
+by requiring an impl to be defined. In can still make sense to explicitly
+implement a named constraint as a way to implement all of the interfaces it
+requires.
 
-### Internal impl
+### Extending an impl
 
-A type that implements an interface _internally_ has all the named members of
-the interface as named members of the type. This means that the members of the
-interface are available by way of both
+A type that _extends_ the implementation an interface has all the named members
```

Comment:
This terminology suggests that there's a point in time where the implementation exists, but hasn't yet been extended, which doesn't match my intuition (or the syntax, where `extend` comes first). Could we instead say the type "extends the interface", or "implements the interface as an extension"?

Reply:
The _syntax_ we decided on is `extend impl as`..., so I'm trying to make the terminology match. We are going to have other kinds of extending (base class, mixin), so whatever terminology we use should hopefully work consistently across those cases.

Reply:
@zygoloid pointed out to me that it is in fact possible (though weird) to implement an interface for a type that is forward declared, and it only extends when the class is later defined.

Reply:
I think you would be able to define the implementation first and extend it later:
```carbon
class A;
impl A as X { fn F() {} }
class A {
  extend impl A as X;
}
```
... though this is certainly not going to be the common case. But I think the implication that defining the implementation and extending it are separate steps is a correct implication.